v030 (#86)
jldec/projectriff.io: _posts/2019-04-09-announcing-riff-0-3-0.md

---
layout: single
title: "Announcing riff v0.3.0"
header:
  overlay_image: /images/frosty-morning.jpg
excerpt:
  With updated Knative components, Windows CLI, and a new Buildpacks design for Functions
permalink: /blog/announcing-riff-0-3-0/
---

We are happy to announce the release of [riff v0.3.0](https://github.com/projectriff/riff/tree/v0.3.x). Thank you all  riff, Buildpacks, and Knative contributors.

The riff CLI can be downloaded from our [releases page](https://github.com/projectriff/riff/releases/tag/v0.3.0) on GitHub. Please follow one of the [getting started](/docs) guides, to create a new cluster on GKE, Minikube, Docker Desktop for Mac, or Docker Desktop for Windows.

Notable changes in this release include:

> - update to Buildpacks 0.1, progress towards user defined invokers/buildpacks

Reviewer: we should mention that the Java buildpack brings in Java 11 so any Maven build POMs should be configured to use Java 8 or later for source version

```
	<properties>
		<maven.compiler.source>1.8</maven.compiler.source>
	</properties>
```

and for Gradle users:

```
sourceCompatibility = 1.8
```

We still care deeply about the potential of stateful stream processing in a polyglot, FaaS environment.

Fortunately, Knative serving is no longer restricted to http/1.1, which unlocks support for streaming protocols like h2, WebSockets, gRPC and RSocket. We have started working on a PoC that includes the following components:

1. Stream CRD with "name" and "provider" (other properties will be added, e.g. "partitions")
2. Processor CRD with "inputs", "outputs", and "function"
3. Stream Gateway providing a generic API that can be backed by different event-log based messaging systems (such as Kafka or Kinesis); initially exploring [liiklus](https://github.com/bsideup/liiklus)
3. Processor will connect to the Stream Gateway(s) for the input/output Streams and interact with the target Function (with which it's colocated in a Pod) via its Invoker layer using gRPC for bidirectional streaming

### More modular Buildpacks
This release introduces a new buildpacks structure with a buildpack per language. 
Here is an updated map of the buildpack-related repos on Github.

(update image, add links)

![](/images/builders.svg)

> - [riff builder](https://github.com/projectriff/riff-buildpack-group) creates the `projectriff/builder` container. 

Reviewer 2: riff-buildpack-group is no longer a thing. It's now called builder

> - [riff buildpack](https://github.com/projectriff/riff-buildpack) contributes invokers for running functions

Reviewer 2: riff-buildpack is no longer a thing. The repo redirects to libfnbuildpack, which is a library for Buildpacks, but not itself a Buildpack.

---
layout: single
title: "Announcing riff v0.3.0"
header:
  overlay_image: /images/frosty-morning.jpg
excerpt:
  With updated Knative components, Windows CLI, and a new Buildpacks design for Functions
permalink: /blog/announcing-riff-0-3-0/
---

We are happy to announce the release of [riff v0.3.0](https://github.com/projectriff/riff/releases/tag/v0.3.0). Thank you all riff, Buildpacks, and Knative contributors.

The riff CLI can be downloaded from our [releases page](https://github.com/projectriff/riff/releases/tag/v0.3.0) on GitHub. Please follow one of the [getting started](/docs) guides, to create a new cluster on GKE, Minikube, Docker Desktop for Mac, or Docker Desktop for Windows.

## Notable changes

### Cloud Native Buildpacks

- update to [pack v0.1.0](https://github.com/buildpack/pack/releases/tag/v0.1.0)
- [builder](https://github.com/projectriff/builder/blob/v0.2.0/builder.toml) uses separate buildpack per function invoker 
- The Java buildpack includes support for Java 11.  Please add a source version of Java 8 or later to your Maven pom or Gradle build.

  #### Maven
  ```xml
  <properties>

>     <maven.compiler.source>1.8</maven.compiler.source>

Reviewer: we need to add target here as well

```
        <maven.compiler.target>1.8</maven.compiler.target>
```

Author: i was wondering about that - how about the gradle side?

Reviewer: nothing additional needed for gradle

- [Java function buildpack](https://github.com/projectriff/java-function-buildpack): contributes [Java invoker](https://github.com/projectriff/java-function-invoker)

#### Node group
- [NodeJS buildpack](https://github.com/cloudfoundry/nodejs-cnb): contributes node.js runtime
- [NPM buildpack](https://github.com/cloudfoundry/npm-cnb): performs npm based builds
- [Node function buildpack](https://github.com/projectriff/node-function-buildpack): contributes the [Node  invoker](https://github.com/projectriff/node-function-invoker) for running JavaScript functions

#### Command group
- [Command function buildpack](https://github.com/projectriff/command-function-buildpack): contributes the [Command invoker](https://github.com/projectriff/command-function-invoker) for running Linux commands.

## Our plans for stream processing

Since we anticipate replacing the existing Channels and Subscriptions with new Stream and Processor resources, aligned with stream-oriented Function Invokers, we are [deprecating](https://github.com/projectriff/riff/pull/1237) the use of Channel and Subscription resources from Knative Eventing in this release.

> You can follow our roadmap for serverless stream processing in [riff issue #1159](https://github.com/projectriff/riff/issues/1159).

Reviewer 2: ```suggestion
You can follow our [roadmap for serverless stream processing](https://github.com/projectriff/riff/issues/1159).
```

permalink: /blog/announcing-riff-0-3-0/
---

We are happy to announce the release of [riff v0.3.0](https://github.com/projectriff/riff/releases/tag/v0.3.0). Thank you all riff, Buildpacks, and Knative contributors.

The riff CLI can be downloaded from our [releases page](https://github.com/projectriff/riff/releases/tag/v0.3.0) on GitHub. Please follow one of the [getting started](/docs) guides, to create a new cluster on GKE, Minikube, Docker Desktop for Mac, or Docker Desktop for Windows.

## Notable changes

### Cloud Native Buildpacks

- update to [pack v0.1.0](https://github.com/buildpack/pack/releases/tag/v0.1.0)
- [builder](https://github.com/projectriff/builder/blob/v0.2.0/builder.toml) uses separate buildpack per function invoker 
- The Java buildpack includes support for Java 11.  Please add a source version of Java 8 or later to your Maven pom or Gradle build.

  #### Maven
  ```xml
  <properties>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
  </properties>
  ```

  #### Gradle
  ```
  sourceCompatibility = 1.8
  ```

### Knative

> - update to [Knative Serving v0.5.1](https://github.com/knative/serving/releases/tag/v0.5.1) and Istio [v1.0.7](https://github.com/istio/istio/releases/tag/1.0.7)

Reviewer: we are using serving 0.5.0 artifacts so maybe change this to `v0.5.0`:

```
Applying resources defined in: https://storage.googleapis.com/knative-releases/serving/previous/v0.5.0/serving.yaml
Applying resources defined in: https://raw.githubusercontent.com/knative/serving/v0.5.0/third_party/config/build/clusterrole.yaml
```

Author: @scothis changed it from 0.5.0 to 0.5.1

---
layout: single
title: "Announcing riff v0.3.0"
header:
  overlay_image: /images/frosty-morning.jpg
excerpt:
  With updated Knative components, Windows CLI, and a new Buildpacks design for Functions
permalink: /blog/announcing-riff-0-3-0/
---

> We are happy to announce the release of [riff v0.3.0](https://github.com/projectriff/riff/tree/v0.3.x). Thank you all  riff, Buildpacks, and Knative contributors.

Reviewer 2: ```suggestion
We are happy to announce the release of [riff v0.3.0](https://github.com/projectriff/riff/releases/tag/v0.3.0). Thank you all riff, Buildpacks, and Knative contributors.
```

The riff CLI can be downloaded from our [releases page](https://github.com/projectriff/riff/releases/tag/v0.3.0) on GitHub. Please follow one of the [getting started](/docs) guides, to create a new cluster on GKE, Minikube, Docker Desktop for Mac, or Docker Desktop for Windows.

## Notable changes

### Cloud Native Buildpacks

- update to [pack v0.1.0](https://github.com/buildpack/pack/releases/tag/v0.1.0)

> - [builder](https://github.com/projectriff/builder/blob/master/builder.toml) uses separate buildpack per function invoker 

Reviewer 2: ```suggestion
- [builder](https://github.com/projectriff/builder/blob/v0.2.0/builder.toml) uses separate buildpack per function invoker 
```

- The Java buildpack includes support for Java 11.  Please add a `compiler.source` to your Maven pom or Gradle config.

  #### Maven
  ```xml
    <properties>
      <maven.compiler.source>1.8</maven.compiler.source>
    </properties>
  ```

  #### Gradle
  ```
  sourceCompatibility = 1.8
  ```

### Knative

> - update to [Knative Serving v0.5.0](https://github.com/knative/serving/releases/tag/v0.5.0) and Istio [v1.0.7](https://github.com/knative/serving/pull/3668/files)

Reviewer 2: ```suggestion
- update to [Knative Serving v0.5.1](https://github.com/knative/serving/releases/tag/v0.5.1) and Istio [v1.0.7](https://github.com/istio/istio/releases/tag/1.0.7)
```

- update to [Knative Build v0.5.0](https://github.com/knative/build/releases/tag/v0.5.0)
- update to [Knative Eventing v0.4.0](https://github.com/knative/eventing/releases/tag/v0.4.0)

### riff CLI

- support for Windows
- easy install via [brew](https://formulae.brew.sh/formula/riff) and [chocolatey](https://chocolatey.org/packages/riff/0.3.0)
- `riff system install` and `riff system uninstall` with improved robustness and cleanup
- `riff namespace init`
    - with basic auth for other registries
- `riff namespace cleanup` similar to system uninstall, but for a namespace
- `riff function build` builds an image without creating a service
- `riff function create`
    - `--sub-path` for `--git-repo` builds from a subdirectory
    - `--image` auto-inferred from registry prefix and function name
    - `riff.toml` optional alternative to CLI parameters
- `riff service invoke` with improved error handling
- `riff channel create` will default to using a default channel provisioner

## More modular Buildpacks

> This release introduces a new buildpacks structure with a buildpack per language. 

Reviewer 2: ```suggestion
This release introduces a new buildpacks structure with a buildpack per language. The modular approach will help restore support for [custom language invokers](https://github.com/projectriff/riff/issues/1093) in a future riff release.
```

- [Command function buildpack](https://github.com/projectriff/command-function-buildpack): contributes the [Command invoker](https://github.com/projectriff/command-function-invoker) for running Linux commands.

## Our plans for eventing
 
In pre-Knative riff we had support for streaming messages between Functions mediated by Topics backed by Kafka. With the transition to Knative, we gave up the ability to stream messages from a Channel to a function and were limited to item-at-a-time request-reply semantics.

We still care deeply about the potential of stateful stream processing in a polyglot, FaaS environment.

Fortunately, Knative serving is no longer restricted to http/1.1, which unlocks support for streaming protocols like h2, WebSockets, gRPC and RSocket. We have started working on a PoC that includes the following components:

1. Stream CRD with "name" and "provider" (other properties will be added, e.g. "partitions")
2. Processor CRD with "inputs", "outputs", and "function"
3. Stream Gateway providing a generic API that can be backed by different event-log based messaging systems (such as Kafka or Kinesis); initially exploring [liiklus](https://github.com/bsideup/liiklus)

> 3. Processor will connect to the Stream Gateway(s) for the input/output Streams and interact with the target Function (with which it's colocated in a Pod) via its Invoker layer using gRPC for bidirectional streaming

Reviewer 2: ```suggestion
4. Processor will connect to the Stream Gateway(s) for the input/output Streams and interact with the target Function (with which it's colocated in a Pod) via its Invoker layer using gRPC for bidirectional streaming
```
